### PR TITLE
Add variadic support for collectAny

### PR DESCRIPTION
**Pull request by poor-circle**
## Why

Finish #129 

## What is changing

add variadic support for collectAny

## Example

now we can write some code like:

```cpp
Try<int> getData()
{
  std::variant<Try<int>,Try<int>, Try<void>> res=collectAny(getDataFromServer1(),getDataFromServer2(),timeout(10s));
  switch (res.index())
  {
    case 0:
    case 1:
       return std::move(std::get<Try<int>>(res));
    case default:
       return Try<int>(std::make_exception_ptr(std::runtime_error("Time out!")))
  }
}

```

### CODE REVIEW

File: async_simple/coro/Collect.h [hunk 1]
```diff
@@ -122,6 +128,83 @@ struct CollectAnyAwaiter {
...
+        await_suspend_impl(std::make_index_sequence<sizeof...(Ts)>{},
+                           std::move(continuation));
+    }
+
+    auto await_resume() {
+        assert(_result != nullptr);
+        return std::move(_result->value());
+    }
+
+    std::unique_ptr<std::tuple<LazyType<Ts>...>> _input;
```
Comment (ChuanqiXu9) [inline async_simple/coro/Collect.h:204]: Is it OK to make `_input` as `std::tuple<LazyType<Ts>...>`?

File: async_simple/coro/Collect.h [hunk 2]
```diff
@@ -122,6 +128,83 @@ struct CollectAnyAwaiter {
     std::shared_ptr<ResultType> _result;
 };
 
+template <template <typename> typename LazyType, typename... Ts>
+struct CollectAnyVariadicAwaiter {
+    using ResultType = std::variant<Try<Ts>...>;
+    using InputType = std::tuple<LazyType<Ts>...>;
+
+    CollectAnyVariadicAwaiter(LazyType<Ts>&&... inputs)
```
Comment (ChuanqiXu9) [inline async_simple/coro/Collect.h:136]: We should remove &&.

File: async_simple/coro/Collect.h [hunk 3]
```diff
@@ -122,6 +128,83 @@ struct CollectAnyAwaiter {
...
+        // Make local copies to shared_ptr to avoid deleting objects too early
+        // if any coroutine finishes before this function.
+        auto result = std::make_shared<std::optional<ResultType>>();
+        auto event =
+            std::make_shared<detail::CountEvent>(std::tuple_size<InputType>());
+
+        _result = result;
+
+        (
+            [&]() {
```
Comment (ChuanqiXu9) [inline async_simple/coro/Collect.h:170]: We'd better align this.
```suggestion
        ( [&]() {
```

File: async_simple/coro/test/LazyTest.cpp [hunk 4]
```diff
@@ -1057,6 +1062,89 @@ TEST_F(LazyTest, testCollectAny) {
...
+        EXPECT_EQ(ret.index(), 2);
+        auto out = std::get<2>(std::move(ret));
+        co_return out;
+    };
+    auto tmp = std::vector<int>{1, 2, 3};
+    auto out = syncAwait(test3().via(&e3));
+    ASSERT_EQ(out.available(), true);
+    ASSERT_EQ(out.value(), tmp);
+
+    std::this_thread::sleep_for(chrono::seconds(2));
```
Comment (ChuanqiXu9) [inline async_simple/coro/test/LazyTest.cpp:1145]: We'd better add a spin lock here.

**Review by qicosmos — APPROVED** [review]
LGTM

File: async_simple/coro/Collect.h (hunk 2 above)
Comment (poor-circle) [reply async_simple/coro/Collect.h:136]: I have removed it at CollectAll()

File: async_simple/coro/Collect.h (hunk 1 above)
Comment (poor-circle) [reply async_simple/coro/Collect.h:204]: move may be expensive for tuple

**Review by ChuanqiXu9 — APPROVED** [review]

**Pull request merged**